### PR TITLE
[ENH] MNE Analyze - application-wide channel selection

### PR DESCRIPTION
**Pull request by gabrielbmotta**
- Channel Selection Plugin
   - controls signal viewer, butterfly view, 2D layout view
- Documentation page for channel selection plugin (see preview [here](https://gabrielbmotta.github.io/mne-cpp/pages/learn/analyze_channelselect.html))
- removed averaging channel selection
- added new event to transmit channel names, channel indices, and channel graphics item parameters

### CODE REVIEW

File: applications/mne_analyze/plugins/channelselection/channelselection.h [hunk 1]
```diff
@@ -158,16 +155,18 @@ private slots:
      */
     void setFiffSettings(QSharedPointer<FIFFLIB::FiffInfo> pFiffInfo);
 
+    Ui::ChannelSelectViewSelect*                        m_pUi;
+
     QPointer<ANSHAREDLIB::Communicator>                 m_pCommu;                   /**< To broadcast signals */
 
-    DISPLIB::SelItem*                                   m_pSelItem;                 /**< Stores parameters from list of QGraphicsItems received by channelselectionview */
+    DISPLIB::SelectionItem*                                   m_pSelItem;                 /**< Stores parameters from list of QGraphicsItems received by channelselectionview */
```
Comment (LorenzE) [inline applications/mne_analyze/plugins/channelselection/channelselection.h:162]: mind indentation

File: applications/mne_analyze/plugins/channelselection/channelselection.cpp [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    if(pNewModel->getType() == MODEL_TYPE::ANSHAREDLIB_FIFFRAW_MODEL) {
+        setFiffSettings(qSharedPointerCast<FiffRawViewModel>(pNewModel)->getFiffInfo());
+    }
+}
+
+//=============================================================================================================
+
+void ChannelSelection::setFiffSettings(QSharedPointer<FIFFLIB::FiffInfo> pFiffInfo)
+{
+    qDebug() << "ChannelSelection::setFiffSettings";
```
Comment (LorenzE) [inline applications/mne_analyze/plugins/channelselection/channelselection.cpp:188]: remove

File: applications/mne_analyze/plugins/channelselection/channelselection.cpp [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+    m_pFiffInfo = pFiffInfo;
+
+    if(m_bIsInit){
+        m_pChannelInfoModel->setFiffInfo(m_pFiffInfo);
+        return;
+    }
+
+    //First time set up only
+
```
Comment (LorenzE) [inline applications/mne_analyze/plugins/channelselection/channelselection.cpp:198]: remove whitespace

File: applications/mne_analyze/plugins/channelselection/channelselection.h [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+     */
+    void selectAll(bool);
+
+    //=========================================================================================================
+    /**
+     * Clear all View Select checkboxes
+     */
+    void selectClear(bool);
+private:
+
```
Comment (LorenzE) [inline applications/mne_analyze/plugins/channelselection/channelselection.h:141]: remove whitespace

File: applications/mne_analyze/plugins/channelselection/channelselection.h [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    /**
+     * Selects all View Select checkboxes
+     */
+    void selectAll(bool);
+
+    //=========================================================================================================
+    /**
+     * Clear all View Select checkboxes
+     */
+    void selectClear(bool);
```
Comment (LorenzE) [inline applications/mne_analyze/plugins/channelselection/channelselection.h:139]: add whitespace

File: applications/mne_analyze/plugins/channelselection/channelselection_global.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
+//=============================================================================================================
+/**
+ * @file     datamanager_global.h
+ * @author   Lorenz Esch <lesch@mgh.harvard.edu>;
+ *           Lars Debor <Lars.Debor@tu-ilmenau.de>;
+ *           Simon Heinke <Simon.Heinke@tu-ilmenau.de>
+ * @since    0.1.0
```
Comment (LorenzE) [inline applications/mne_analyze/plugins/channelselection/channelselection_global.h:7]: 01.5

File: applications/mne_analyze/plugins/channelselection/channelselection_global.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
+//=============================================================================================================
+/**
+ * @file     datamanager_global.h
```
Comment (LorenzE) [inline applications/mne_analyze/plugins/channelselection/channelselection_global.h:3]: rename

File: applications/mne_analyze/plugins/channelselection/channelselection_global.h [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+ * @author   Lorenz Esch <lesch@mgh.harvard.edu>;
```
Comment (LorenzE) [inline applications/mne_analyze/plugins/channelselection/channelselection_global.h:4]: update authors

File: doc/gh-pages/pages/learn/analyze_channelselect.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+This plugin lets you select which channels will be visible in the views of MNE Analyze.
+
+![](../../images/analyze/mne_an_chanselect_1.png)
+
+Using the Channel Selection View, you can manually select which channels will be displayed, by either clicking individual channels or dragging a selection box around them.
+For this to work correctly, make sure the correct layout is chosen in the dropdown at the bottom, so that the view corresponds to the channels in the acquisition device used.
+
+![](../../images/analyze/mne_an_chanselect_2.png)
+
+From the controls you can select from premade selection groups, or define your won. You can also select which views the channel selection will be applied to from the list at the bottom
```
Comment (LorenzE) [inline doc/gh-pages/pages/learn/analyze_channelselect.md:19]: "your own" "list at the bottom."

File: doc/gh-pages/pages/learn/analyze_channelselect.md [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+nav_order: 7
+---
+# Channel Selection
+
+This plugin lets you select which channels will be visible in the views of MNE Analyze.
+
+![](../../images/analyze/mne_an_chanselect_1.png)
+
+Using the Channel Selection View, you can manually select which channels will be displayed, by either clicking individual channels or dragging a selection box around them.
+For this to work correctly, make sure the correct layout is chosen in the dropdown at the bottom, so that the view corresponds to the channels in the acquisition device used.
```
Comment (LorenzE) [inline doc/gh-pages/pages/learn/analyze_channelselect.md:15]: "drop down at the bottom" should now be "at the top"

File: libraries/disp/disp.pro [hunk 11]
```diff
@@ -37,7 +37,7 @@ include(../../mne-cpp.pri)
 
 TEMPLATE = lib
 
-CONFIG += skip_target_version_ext
+CONFIG += skip_target_version_ext console
```
Comment (LorenzE) [inline libraries/disp/disp.pro:40]: do we need the console flag?

File: libraries/disp/viewers/averagelayoutview.cpp [hunk 12]
```diff
@@ -76,6 +77,7 @@ AverageLayoutView::AverageLayoutView(const QString& sSettingsPath,
                                      QWidget *parent,
                                      Qt::WindowFlags f)
 : AbstractView(parent, f)
+, m_pFiffInfo(Q_NULLPTR)
```
Comment (LorenzE) [inline libraries/disp/viewers/averagelayoutview.cpp:80]: QSharedPointers are always initialized as NULL so this is not needed here.

File: libraries/disp/viewers/averagelayoutview.cpp [hunk 13]
```diff
@@ -269,11 +289,45 @@ void AverageLayoutView::channelSelectionManagerChanged(const QList<QGraphicsItem
 
 void AverageLayoutView::updateData()
 {
+    if(m_pFiffInfo) {
+
```
Comment (LorenzE) [inline libraries/disp/viewers/averagelayoutview.cpp:293]: remove whitespace

File: libraries/disp/viewers/averagelayoutview.h [hunk 14]
```diff
@@ -239,8 +256,12 @@ class DISPSHARED_EXPORT AverageLayoutView : public AbstractView
     QSharedPointer<DISPLIB::EvokedSetModel>                     m_pEvokedSetModel;          /**< The data model */
     QSharedPointer<DISPLIB::ChannelInfoModel>                   m_pChannelInfoModel;        /**< Channel info model. */
 
+    QSharedPointer<FIFFLIB::FiffInfo>                           m_pFiffInfo;                /**< FiffInfo for currently loaded file */
+
     QPointer<QGraphicsView>                                     m_pAverageLayoutView;       /**< View for 2D average layout scene */
 
+    SelectionItem*                                              m_pSelItem;
```
Comment (LorenzE) [inline libraries/disp/viewers/averagelayoutview.h:263]: are you making sure that these are either passed to a parent/child relationship or deleted manually?

File: libraries/disp/viewers/butterflyview.h [hunk 15]
```diff
@@ -236,6 +236,9 @@ class DISPSHARED_EXPORT ButterflyView : public QWidget
      */
     void showSelectedChannelsOnly(const QStringList& selectedChannels);
 
+    //=========================================================================================================
```
Comment (LorenzE) [inline libraries/disp/viewers/butterflyview.h:239]: add docu

File: libraries/disp/viewers/helpers/averagescene.cpp [hunk 16]
```diff
@@ -106,6 +107,26 @@ void AverageScene::repaintItems(const QList<QGraphicsItem *> &selectedChannelIte
 
 //=============================================================================================================
 
+void AverageScene::repaintSelItems(const DISPLIB::SelectionItem &selectedChannelItems)
```
Comment (LorenzE) [inline libraries/disp/viewers/helpers/averagescene.cpp:110]: rename to `repaintSelctionItems`

File: libraries/disp/viewers/helpers/selectionsceneitem.cpp [hunk 17]
```diff
@@ -143,3 +143,12 @@ void SelectionSceneItem::paint(QPainter *painter,
     painter->drawStaticText(-15+((30-sizeText.width())/2), -32, staticElectrodeName);
 }
 
+//=============================================================================================================
+
+SelectionItem::SelectionItem(const SelectionItem &other){
```
Comment (LorenzE) [inline libraries/disp/viewers/helpers/selectionsceneitem.cpp:148]: add new line between ) and {

File: libraries/disp/viewers/averagelayoutview.h (hunk 14 above)
Comment (gabrielbmotta) [reply libraries/disp/viewers/averagelayoutview.h]: Removed as a member variable

**Pull request merged**